include/knowhere/index/index_node.h

```diff
@@ -540,46 +571,102 @@ class IndexIterator : public IndexNode::iterator {
 
     bool initialized_ = false;
     bool retain_iterator_order_ = false;
+    bool use_knowhere_search_pool_ = true;
     const int64_t sign_;
 };
 
-// An iterator implementation that accepts a list of distances and ids and returns them in order.
+// An iterator implementation that accepts a function to get distances and ids list and returns them in order.
+// We do not directly accept a distance list as input. The main reason for this is to minimize the initialization time
+//   for all types of iterators in the `ANNIterator` interface, moving heavy computations to the first '->Next()' call.
+//   This way, the iterator initialization does not need to perform any concurrent acceleration, and the search pool
+//   only needs to handle the heavy work of `->Next()`
+template <typename Compute_Dist_Func>
 class PrecomputedDistanceIterator : public IndexNode::iterator {
  public:
-    PrecomputedDistanceIterator(std::vector<DistId>&& distances_ids, bool larger_is_closer)
-        : larger_is_closer_(larger_is_closer), results_(std::move(distances_ids)) {
-        sort_size_ = get_sort_size(results_.size());
-        sort_next();
-    }
-
-    // Construct an iterator from a list of distances with index being id, filtering out zero distances.
-    PrecomputedDistanceIterator(const std::vector<float>& distances, bool larger_is_closer)
-        : larger_is_closer_(larger_is_closer) {
-        // 30% is a ratio guesstimate of non-zero distances: probability of 2 random sparse splade vectors(100 non zero
-        // dims out of 30000 total dims) sharing at least 1 common non-zero dimension.
-        results_.reserve(distances.size() * 0.3);
-        for (size_t i = 0; i < distances.size(); i++) {
-            if (distances[i] != 0) {
```

Comment:
Why we do this? I think this `PrecomputedDistanceIterator` is used for both dense/sparse BF
@zhengbuqian

Reply:
submitted a new commit to reduce misunderstanding.